[DataPipe] Fix error message coming from single iterator constraint

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #79547

Fixes https://github.com/pytorch/data/issues/516

* Prevent confusion by not adding the extra string about `__iter__` when the exception is raised because of single iterator constraint
* Added a missing space
* Make sure `__len__` exists before calling it during exception handling
* Handles the case where there was no initial exception message (previously skipped)


Exception example:
```python
from torchdata.datapipes.iter import IterableWrapper

source_dp = IterableWrapper(range(10))
it1 = iter(source_dp)
next(it1)
it2 = iter(source_dp)
next(it2)
next(it1)
```
Will raise:
```
Traceback (most recent call last):
  File "/Users/scratch/fast_forward.py", line 32, in <module>
    next(it1)
  File "/Users/pytorch/torch/utils/data/datapipes/_typing.py", line 524, in wrap_generator
    _check_iterator_valid(datapipe, iterator_id)
  File "/Users/pytorch/torch/utils/data/datapipes/_typing.py", line 445, in _check_iterator_valid
    raise RuntimeError(_gen_invalid_iterdatapipe_msg(datapipe) + _feedback_msg)
RuntimeError: This iterator has been invalidated because another iterator has been created from the same IterDataPipe: IterableWrapperIterDataPipe(deepcopy=True, iterable=range(0, 10))
This may be caused multiple references to the same IterDataPipe. We recommend using `.fork()` if that is necessary.
For feedback regarding this single iterator per IterDataPipe constraint, feel free to comment on this issue: https://github.com/pytorch/data/issues/45.
```

NivekT committed:
[DataPipe] Fix error message coming from singler iterator constraint

[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -530,8 +530,10 @@ def wrap_generator(*args, **kwargs):
-                full_msg = f"{msg} {datapipe.__class__.__name__}({_generate_input_args_string(datapipe)})"
-                if len(e.args) >= 1 and msg not in e.args[0]:
+                single_iterator_msg = "single iterator per IterDataPipe constraint"
+                has_len = hasattr(e.args, '__len__') and len(e.args) >= 1
```
Noob question: What would be the cause of this Error with `__len__`? Someone calling `len(dp)` in the middle of iteration?

NivekT replied:
The extra logic here is not related to the DataPipe,but rather, the exception being raised.

It is checking if `e.args` (arguments of the exception) has `__len__` and if it does, there is more than one element. I have seen cases where `e.args` does not mean that criteria.

ejguan replied:
I see. Thanks for the explanation.

NivekT replied:
Your question prompted me to take another look and handle the case where `len(e.args) == 0` when there is no exception message. Have a look.

NivekT replied:
```python
class MyDataPipe(IterDataPipe):

    def __init__(self, x):
        self.x = x

    def __iter__(self):
        raise RuntimeError
        yield 0


dp = MyDataPipe(1)
it = iter(dp)
next(it)
```

Previously, it will just show:
```
Traceback (most recent call last):
  File "/Users/scratch/graphvisualization.py", line 27, in <module>
    next(it)
  File "/Users/pytorch/torch/utils/data/datapipes/_typing.py", line 514, in wrap_generator
    response = gen.send(None)
  File "/Users/scratch/graphvisualization.py", line 21, in __iter__
    raise RuntimeError
RuntimeError: 
```
Now it adds the extra message in the end:
```
This exception is thrown by __iter__ of MyDataPipe(x=1)
```

ejguan replied:
Makes sense. I hope at least TorchData doesn't have such kind of Error without message. lol

ejguan reviewed (approved):
LGTM

NivekT committed:
Update on "[DataPipe] Fix error message coming from single iterator constraint"


Fixes https://github.com/pytorch/data/issues/516

* Prevent confusion by not adding the extra string about `__iter__` when the exception is raised because of single iterator constraint
* Added a missing space
* Make sure `__len__` exists before calling it during exception handling


Exception example:
```python
from torchdata.datapipes.iter import IterableWrapper

source_dp = IterableWrapper(range(10))
it1 = iter(source_dp)
next(it1)
it2 = iter(source_dp)
next(it2)
next(it1)
```
Will raise:
```
Traceback (most recent call last):
  File "/Users/scratch/fast_forward.py", line 32, in <module>
    next(it1)
  File "/Users/pytorch/torch/utils/data/datapipes/_typing.py", line 524, in wrap_generator
    _check_iterator_valid(datapipe, iterator_id)
  File "/Users/pytorch/torch/utils/data/datapipes/_typing.py", line 445, in _check_iterator_valid
    raise RuntimeError(_gen_invalid_iterdatapipe_msg(datapipe) + _feedback_msg)
RuntimeError: This iterator has been invalidated because another iterator has been created from the same IterDataPipe: IterableWrapperIterDataPipe(deepcopy=True, iterable=range(0, 10))
This may be caused multiple references to the same IterDataPipe. We recommend using `.fork()` if that is necessary.
For feedback regarding this single iterator per IterDataPipe constraint, feel free to comment on this issue: https://github.com/pytorch/data/issues/45.
```

[ghstack-poisoned]

NivekT commented:
@pytorchbot merge
